### PR TITLE
fix(ivy): handling className as an input properly

### PR DESCRIPTION
**Pull request by AndrewKushnir**
Prior to this commit, all `className` inputs were not set because the runtime code assumed that the `classMap` instruction is only generated for `[class]` bindings. However the `[className]` binding also produces the same `classMap`, thus the code needs to distinguish between `class` and `className`. This commit adds extra logic to select the right input name and also throws an error in case `[class]` and `[className]` bindings are used on the same element simultaneously.

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/render3/instructions/element.ts [hunk 1]
```diff
@@ -132,7 +132,8 @@ export function ɵɵelementEnd(): void {
   }
 
   if (hasClassInput(tNode)) {
-    setDirectiveStylingInput(tNode.classes, lView, tNode.inputs !['class']);
+    const inputName = tNode.inputs !.hasOwnProperty('class') ? 'class' : 'className';
```
Comment (matsko) [inline packages/core/src/render3/instructions/element.ts]: Can you add a todo:

```
TODO (matsko|AndrewKushnir): refactor this once we figure out how to generate separate `input('class') + classMap()` instructions
```

File: packages/core/src/render3/instructions/styling.ts [hunk 2]
```diff
@@ -402,7 +402,10 @@ function updateDirectiveInputValue(
     // directive input(s) in the event that it is falsy during the
     // first update pass.
     if (newValue || isContextLocked(context, false)) {
-      const inputName = isClassBased ? 'class' : 'style';
+      let inputName: string = 'style';
+      if (isClassBased) {
+        inputName = tNode.inputs !.hasOwnProperty('class') ? 'class' : 'className';
```
Comment (matsko) [inline packages/core/src/render3/instructions/styling.ts]: This is also used in `element.ts`. Could you create a shared function inside of `styling_util.ts` that figures out what inputName to provide? Then we could add the TODO I mentioned above in that function.

**Review by matsko — APPROVED** [review]
Approved. With a simple function change.

File: packages/core/test/acceptance/styling_spec.ts [hunk 3]
```diff
@@ -551,6 +551,52 @@ describe('styling', () => {
...
+  it('should write to a `className` input binding', () => {
+    @Component({
+      selector: 'comp',
+      template: `{{className}}`,
+    })
+    class Comp {
+      @Input() className: string = '';
+    }
+    @Component({
+      template: `<comp [className]="'my-className'"></comp>`,
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/styling_spec.ts]: could we have another test for the case of a static class (`<comp class="static" [className]="'my-className'"></comp>`)? It is not clear for me if this situation will be dissolved as well?

Comment (AndrewKushnir) [reply packages/core/test/acceptance/styling_spec.ts]: I've added extra test to make sure that the use-case that you've mentioned is covered. The only difference is that in Ivy, writing to the `className` input binding has a side-effect of applying styling (i.e. static `class` value is combined with `className` binding value). Thank you.

**Comment by AndrewKushnir** [comment]
Thanks @pkozlowski-opensource and @matsko for review. I've performed additional changes based on your comments. Could you please have a look? Thank you.

**Review by matsko — APPROVED** [review]

File: packages/compiler/src/render3/view/styling_builder.ts [hunk 4]
```diff
@@ -201,6 +201,10 @@ export class StylingBuilder {
     const entry:
         BoundStylingEntry = {name: property, value, sourceSpan, hasOverrideFlag, unit: null};
     if (isMapBased) {
+      if (this._classMapInput) {
+        throw new Error(
+            '[class] and [className] bindings can not be used on the same element simultaneously');
```
Comment (kara) [inline packages/compiler/src/render3/view/styling_builder.ts]:
```suggestion
            '[class] and [className] bindings cannot be used on the same element simultaneously');
```

Comment (AndrewKushnir) [reply packages/compiler/src/render3/view/styling_builder.ts]: Thanks Kara!

**Comment by AndrewKushnir** [comment]
- [VE Presubmit](http://test/OCL:275146947:BASE:275146983:1571270984207:b6508193)
- [Ivy Presubmit](http://test/OCL:275146947:BASE:275147104:1571271032616:34acf646)

**Pull request closed without merging**